### PR TITLE
i190: Add a nice UI during app initialization

### LINKED ISSUE #190: Make sure initialization on app load presents a nice UI
#### Is your feature request related to a problem? Please describe.
Currently on app load, there is a CircularProgressIndicator while waiting, but it looks terrible as it's not inside the MaterialApp widget.  There is also no display of any error that occurs during initialisation of firebase and the redux store. 

#### Describe the solution you'd like
While waiting show a nice CircularProgressIndicator and a Text with status, eg. "Initializing Firebase" and "Initializing the redux store".  

If there is an error display in an alert. 
- we can't use the redux store yet so we'll need a separate function somewhere that displays an alert with the error

### PR DESCRIPTION
**Pull request by nickmeinhold**
fixes #190

### CODE REVIEW

File: lib/widgets/auth/auth_page.dart [hunk 1]
```diff
@@ -44,21 +45,6 @@ class _AuthPageState extends State<AuthPage> {
   }
 }
 
-class WaitingIndicator extends StatelessWidget {
```
Comment (nickmeinhold) [inline lib/widgets/auth/auth_page.dart:47]: this widget was moved to it's own file in the 'widgets/shared' folder

File: lib/widgets/crowd_league_app.dart [hunk 2]
```diff
@@ -122,3 +123,64 @@ class CheckAuth extends StatelessWidget {
         });
   }
 }
+
+class InitializingIndicator extends StatelessWidget {
```
Comment (nickmeinhold) [inline lib/widgets/crowd_league_app.dart:127]: This widget is just a CircularProgressIndicator and some text, in a Material widget so it looks nice.  It's a separate widget as the existing ProgressIndicator widget doesn't need to have it's contents in a Material widget.

File: lib/widgets/crowd_league_app.dart [hunk 3]
```diff
@@ -122,3 +123,64 @@ class CheckAuth extends StatelessWidget {
...
+          children: <Widget>[
+            CircularProgressIndicator(),
+            SizedBox(height: 30),
+            Text(message, textDirection: TextDirection.ltr)
+          ]),
+    );
+  }
+}
+
+class ErrorPage extends StatelessWidget {
```
Comment (nickmeinhold) [inline lib/widgets/crowd_league_app.dart:158]: This widget just displays the available info if there is an error during intialization.  It's not particularly pretty but it shouldn't ever be seen and if it is, we just need to see the info so I think it's fine.

File: lib/widgets/shared/waiting_indicator.dart [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import 'package:flutter/material.dart';
```
Comment (nickmeinhold) [inline lib/widgets/shared/waiting_indicator.dart:1]: moved out of `auth_page.dart`

File: lib/widgets/shared/waiting_indicator.dart [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+class WaitingIndicator extends StatelessWidget {
```
Comment (nickmeinhold) [inline lib/widgets/shared/waiting_indicator.dart:3]: moved out of `auth_page.dart`

File: lib/widgets/crowd_league_app.dart (hunk 2 above)
Comment (tamari-gray) [reply lib/widgets/crowd_league_app.dart:127]: thanks for the detailed comments :D

File: lib/widgets/shared/waiting_indicator.dart (hunk 5 above)
Comment (tamari-gray) [reply lib/widgets/shared/waiting_indicator.dart:3]: 🙏

**Review by tamari-gray — DISMISSED** [review]
looks good to me mate! 
i had a breezy review with thos crystal clear comments 🔮 
would you be keen to try the pr bonus stage and add some tests? 🎁 🎁 ( cue nintendo dance music )

**Comment by nickmeinhold** [comment]
Thanks so much @tamari-gray! Really appreciate the challenge to add some tests - so glad we did this, it meant setting up the CrowdLeagueApp widget to allow injecting, using a wrapper around the Firebase initialisation static function, making fakes with completers... and I swear there were a couple more tricks in there too!  Awesome to have that setup though, will make testing anything added to the very top of our widget tree nice and easy to test!

File: test/mocks/wrappers/firebase_wrapper_mocks.dart [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import 'dart:async';
+
+import 'package:crowdleague/utils/wrappers/firebase_wrapper.dart';
+import 'package:firebase_core/firebase_core.dart';
+
+class FakeFirebaseWrapper implements FirebaseWrapper {
```
Comment (tamari-gray) [inline test/mocks/wrappers/firebase_wrapper_mocks.dart:6]: is this meant to be a mockito fake?

Comment (nickmeinhold) [reply test/mocks/wrappers/firebase_wrapper_mocks.dart:6]: thanks mate!  great question!  it wasn't meant to use mockito in that I just wanted to have all the overridden methods there and throw an Unimplemented error, so I just add implements and have the autocorrect fill them out then as the tests break I add a method implementation.  It's more chaos than by design - should it be named differently do you think?  is there a reason to use a mockito fake over a 'real' object that is (as far as I understand) essentially the same thing...?

Comment (nickmeinhold) [reply test/mocks/wrappers/firebase_wrapper_mocks.dart:6]: oh by the way I meant to ask if you wanted me to break up the PR, it turned into a bit of a shit show... my 3am coding brain was making some bad calls I think and I threw a bunch of crap in here... I'll go through and add some comments now 🙂

Comment (nickmeinhold) [reply test/mocks/wrappers/firebase_wrapper_mocks.dart:6]: and very happy to pull it apart if you'd like, I wanna play around with using cherry picks to split up a PR anyway

Comment (tamari-gray) [reply test/mocks/wrappers/firebase_wrapper_mocks.dart:6]: ooh yeah if your happy to, pulling it apart would help alot 😄

Comment (tamari-gray) [reply test/mocks/wrappers/firebase_wrapper_mocks.dart:6]: and i think naming is fine as is :) and if u want to use all methods then hell yeah we dont need mockito then!

Comment (tamari-gray) [reply test/mocks/wrappers/firebase_wrapper_mocks.dart:6]: shell be right, thanks for the work!

File: lib/widgets/crowd_league_app.dart [hunk 7]
```diff
@@ -9,41 +9,51 @@ import 'package:crowdleague/models/app/settings.dart';
...
+  CrowdLeagueApp({FirebaseWrapper firebase, ServicesBundle redux})
+      : _firebase = firebase ?? FirebaseWrapper(),
+        _redux = redux;
   @override
   _CrowdLeagueAppState createState() => _CrowdLeagueAppState();
 }
 
 class _CrowdLeagueAppState extends State<CrowdLeagueApp> {
-  final _navKey = GlobalKey<NavigatorState>();
+  ServicesBundle _redux;
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply lib/widgets/crowd_league_app.dart:35]: 😮 beautiful!!
(comment helps so much thank you 🙏 )

File: test/mocks/wrappers/firebase_wrapper_mocks.dart [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  final Completer<FirebaseApp> _completer;
+
+  FakeFirebaseWrapper({Completer<FirebaseApp> completer})
+      : _completer = completer ?? Completer<FirebaseApp>();
+
+  /// Returns a [Future] that completes when the supplied [Completer] says so
```
Comment (tamari-gray) [inline test/mocks/wrappers/firebase_wrapper_mocks.dart:12]: you will complete, or i will make you complete, your in my world now future

198b4daec66737f1010627b9f1a1e0a2_400x400

File: test/utils/services_bundle_mocks.dart [hunk 9]
```diff
@@ -0,0 +1,63 @@
+import 'dart:async';
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/utils/services_bundle_mocks.dart:1]: cant handle the truth 😆

File: test/widget/crowd_league_app/crowd_league_app_test.dart [hunk 10]
```diff
@@ -0,0 +1,93 @@
+import 'dart:async';
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/widget/crowd_league_app/crowd_league_app_test.dart:1]: oooh been looking forward to reading these, thanks for the heads up!

File: test/mocks/wrappers/firebase_wrapper_mocks.dart (hunk 8 above)
Comment (nickmeinhold) [reply test/mocks/wrappers/firebase_wrapper_mocks.dart:12]: Hahahaha with just emoji your review comments are damn funny, the memes are gonna ruin me 🤣

File: test/widget/crowd_league_app/crowd_league_app_test.dart [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+      final initializingIndicatorFinder = find.byType(InitializingIndicator);
+
+      // verify that the InitializingIndicator is shown
+      expect(initializingIndicatorFinder, findsOneWidget);
+
+      // verify the expected text is shown, indicating waiting for both
+      expect(find.text('Waiting for Firebase...'), findsOneWidget);
+
+      // complete the firebase future and verfiy text has changed as expected
+      firebaseCompleter.complete();
```
Comment (tamari-gray) [inline test/widget/crowd_league_app/crowd_league_app_test.dart:47]: all the setup was well worth it. thanks for pushing through!!

Comment (nickmeinhold) [reply test/widget/crowd_league_app/crowd_league_app_test.dart:47]: Yeah good to have figured it out I reckon, should be real helpful down the track I'd say

Comment (nickmeinhold) [reply test/widget/crowd_league_app/crowd_league_app_test.dart:47]: Thanks mate!

**Review by tamari-gray — APPROVED** [review]
what can i say, 
detailed comments to help me **easily** understand the pr and some beautiful tests.
amazing work boss 🥳

File: lib/utils/wrappers/firebase_wrapper.dart [hunk 12]
```diff
@@ -0,0 +1,7 @@
+import 'package:firebase_core/firebase_core.dart';
+
+/// Wraps the [Firebase] static function so we can optionally inject a mock in
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply lib/utils/wrappers/firebase_wrapper.dart:3]: i think so :)

**Comment by nickmeinhold** [comment]
Awesome team effort!  Really stoked to have gotten the test double naming thing sorted!  I made an issue to do that separately - thanks for asking great questions!

**Pull request merged**